### PR TITLE
Stacked Utils refactor

### LINKED ISSUE #2262: Stacked Utils Refactor
- [x] StackedUtils’s computeStackOffsets() should not return a d3Map and likely a Utils.Map.
- [x] StackedUtil’s computeStackExtents() has quite the number of arguments (5). Seek to reduce this.
- [x] `Utils.Stacked` -> `Utils.Stacking`

### PR DESCRIPTION
**Pull request by aicioara**
Closes #2262

DONE:
- No more exposing of `d3.Map()`
- Stacked is now a module, instead of a class
- `Utils.Stacked` -> `Utils.Stacking`
- `Utils.StackedDatum` -> `Utils.Stacking.StackedDatum`
- `computeStackExtent` takes 3 arguments instead of 5
- `computeStackExtent()` -> `stackedExtent()`
- `computeStackOffsets()` -> `stack()`
- created `Utils.Stacking.StackingResult` exported type
- added better documentation

API Breaking changes:
- `computeStackOffsets()` -> `stack()`
- `computeStackExtent(datasets, keyAccessor, valueAccessor, stackOffsets, filter)` -> `computeStackExtent(stackOffsets, keyAccessor, filter)`
- `computeStackExtent()` -> `stackedExtent()`
- created `Utils.Stacking.StackingResult` exported type
- created `Utils.Stacking.StackedDatum` exported type
- `domainKeys()` -> DELETED (privatized) 

TODO: 
- <del> The hacky method `XYPlot._makeFilterByProperty` </del>
- <del> Rename `Utils.Stacked` -> `Utils.Stacking` </del>
- <del> Do not use `String()` everywhere in the `StackedAreaPlot` and `StackedBarPlot` for getting things inside the `_stackOffsets` inner map </del>
- <del> rename `_stackOffsets` to something more sensible as we are now returning more data </del>
- <del> as per above, rename `Utils.Stacked.computeStackOffsets` to something more sensible </del>

**QE Notes:**
Stacked Area and Stacked Bar should be affected. It should just be a regression pass, no functionality affected. Try stacking data that mismatches for bar plot, or data where keys are weird

### CODE REVIEW

File: plottable.d.ts [hunk 1]
```diff
@@ -232,26 +232,35 @@ declare module Plottable {
...
              * for drawing purposes.
              *
              * @return {Utils.Map<Dataset, d3.Map<number>>} A map from each dataset to the offset of each datapoint
              */
-            static computeStackOffsets(datasets: Dataset[], keyAccessor: Accessor<any>, valueAccessor: Accessor<number>): Map<Dataset, d3.Map<number>>;
+            function computeStackOffsets(datasets: Dataset[], keyAccessor: Accessor<any>, valueAccessor: Accessor<number>): Map<Dataset, Map<string, {
+                key: any;
+                value: number;
+                offset?: number;
+            }>>;
```
Comment (bluong) [inline plottable.d.ts]: Probably need to explicitly type `StackedDatum` here

**Comment by jtlan** [comment]
There might not be a good way to get rid of the `string`-ification unless we have a good way to check for `null` and `undefined` keys.

**Comment by aicioara** [comment]
`computeStackOffsets()` -> `computeStackInformation()` ?

**Comment by aicioara** [comment]
Still working on this

**Comment by aicioara** [comment]
I'd also break `computeStackExtent` into 2 methods. Right now it is doing too much. Why does the computation of the stack extent need to know about the filter?

File: src/utils/stackingUtils.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        value: number;
+        offset: number;
+      };
+
+      export type StackInformation = Utils.Map<Dataset, Utils.Map<string, StackedDatum>>;
+
+      var nativeMath: Math = (<any>window).Math;
+
+      /**
+       * Computes the stacking information (value and offset) for each data point in each dataset
```
Comment (jtlan) [inline src/utils/stackingUtils.ts]: "Dataset" (capitalized)

File: src/utils/stackingUtils.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+              offset: offset
+            });
+          });
+          datasetToKeyToStackedDatum.set(dataset, keyToStackedDatum);
+        });
+        return datasetToKeyToStackedDatum;
+      }
+
+      /**
+       * Computes the total extent over all data points in all datasets, taking stacking into consideration
```
Comment (jtlan) [inline src/utils/stackingUtils.ts]: "Datasets" (capitalized)

File: src/utils/stackingUtils.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        });
+        var maxStackExtent = Utils.Math.max(extents, 0);
+        var minStackExtent = Utils.Math.min(extents, 0);
+
+        return [nativeMath.min(minStackExtent, 0), nativeMath.max(0, maxStackExtent)];
+      }
+
+      /**
+       * Given an array of datasets and the accessor function for the key, computes the
+       * set reunion (no duplicates) of the domain of each dataset.
```
Comment (jtlan) [inline src/utils/stackingUtils.ts]: Capitalization here and line above

**Comment by jtlan** [comment]
>Why does the computation of the stack extent need to know about the filter?

It needs to know which are valid values to include when computing the extent. Compare `_computeExtent()` in `Plot`.

File: src/utils/stackingUtils.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+       */
+      export function domainKeys(datasets: Dataset[], keyAccessor: Accessor<any>) {
+        var domainKeys = d3.set();
```
Comment (jtlan) [inline src/utils/stackingUtils.ts]: I'll note that this method normalizes the keys by stringifying them, but there's no warning of this for the user.

File: src/utils/stackingUtils.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+module Plottable {
+  export module Utils {
+    export module Stacking {
+
+      export type StackedDatum = {
+        value: number;
+        offset: number;
+      };
+
+      export type StackInformation = Utils.Map<Dataset, Utils.Map<string, StackedDatum>>;
```
Comment (jtlan) [inline src/utils/stackingUtils.ts]: "`DatasetOffsetMap`"?

File: src/utils/stackingUtils.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+          dataset.data().forEach((datum, index) => {
+            domainKeys.add(keyAccessor(datum, index, dataset));
+          });
+        });
+
+        return domainKeys.values();
+      }
+
+      /**
+       * Normalizes a key to be used when for stacking
```
Comment (jtlan) [inline src/utils/stackingUtils.ts]: "Returns a normalized key used for stacking."

File: src/utils/stackingUtils.ts (hunk 6 above)
Comment (aicioara) [inline src/utils/stackingUtils.ts]: It's not just offset, right?

Comment (jtlan) [inline src/utils/stackingUtils.ts]: "`DatasetToStackResult`"?

Comment (jtlan) [inline src/utils/stackingUtils.ts]: From offline: `StackingResult`

**Comment by aicioara** [comment]
Pushed again. Updating the list of changes

File: src/utils/stackingUtils.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+      /**
+       * Computes the total extent over all data points in all Datasets, taking stacking into consideration.
+       *
+       * @param {StackingResult} stackInformation The value and offset information for each datapoint in each dataset
+       * @oaram {Accessor<any>} keyAccessor Accessor for the key of the data existent in the stackInformation
+       * @param {Accessor<boolean>} filter A filter for data to be considered when computing the total extent
+       * @return {[number, number]} The total extent
+       */
+      export function stackedExtent(stackInformation: StackingResult, keyAccessor: Accessor<any>, filter: Accessor<boolean>) {
```
Comment (jtlan) [inline src/utils/stackingUtils.ts]: param name should match the type `StackingResult`

**Comment by crmorford** [comment]
What are these for?
created Utils.Stacking.StackingResult exported type
created Utils.Stacking.StackedDatum exported type

**Comment by jtlan** [comment]
>What are these for?
created Utils.Stacking.StackingResult exported type
created Utils.Stacking.StackedDatum exported type

They already existed, but were renamed. Essentially they're used to hold the result of the `stack()` method, which computes offsets for stacked `Dataset`s.

**Comment by aicioara** [comment]
I merged in an overlaying test fix from develop

**Comment by aicioara** [comment]
About the exported types:
StackingDatum existed before, it was not exported. StackinResult existed before, it was called `Map<Dataset, Map<string, Utils.StackingDatum>>`

**Comment by crmorford** [comment]
The symbols quicktest is still broken, but looks good otherwise.

**Pull request merged**